### PR TITLE
RFC: manifests

### PR DESCRIPTION
**Pull request by b5**

### CODE REVIEW

File: text/0019-manifests.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+[summary]: #summary
+
+Define dataset manifests: a building block for managing subsets of a dataset and point-to-point block syncing.
+
+# Motivation
+[motivation]: #motivation
+
+At Qri we've run into a pretty serious snag: sending a datasets from point-to-point in a manner that's roughly comparable with a centralized service in terms of speed. While working on `qri publish`, we want to pin the blocks in a dataset on an IPFS node that is always available so others can get to it. This requires getting the dataset from a local machine to a publically-hosted Qri registry. This is the rough equivalent of `git push origin master` for Qri.
+
+Without the Qri verion of `git push origin master`, Qri's usefulness is degraded. We've written a test that uses out-of-band communication to ask a remote node to `ipfs pin add` while the local machine is on the distribted web, and it's taking anywhere from 5-7 minutes to pin a ~200kb dataset, even with an explicit request to directly connect the two nodes making the transfer.
```
Comment (lightandluck) [inline text/0019-manifests.md]: typos: verion, distribted

File: text/0019-manifests.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+While we're most certianly doing a few things wrong in terms of network configuration, this "ambient pinning" approach isn't producing a consistent-enough user experience. After initially [proposing some upstream changes](https://github.com/ipld/specs/pull/66) in IPFS, it's become clear that it'll be more useful to others to solve our own pain points with an eye toward generalization. If we come up with a solution that works well, it might be useful in other situations.
```
Comment (lightandluck) [inline text/0019-manifests.md]: typo: certianly

File: text/0019-manifests.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+* comparing blocks of a dataset between two versions to examine deduplication
+
+Because datasets in Qri are organized into _immutable graphs_ of content addressed blocks. There are two versions of what it means to "have" a dataset. In the first sense of the word, you have the entire graph of blocks that make up a dataset pinned locally. The other meaning of "having" is the result of reading the canonical hash and encoding them into an API, CLI, or exported dataset.
+
+This second sense of "having" is obviously vital for Qri to work properly because the entire world doesn't read raw IPFS blocks, and never will. The problem is the first version is all-or-nothing. You either have the _entire_ dataset graph, or you have to consider the result an emphemeral biproduct of the canonical source.
+
+Before we approach the point-to-point sync problem, I'd like to introduce **Manifests** as a method of describing all blocks that compose a dataset in a concise, repeatable way. By having a clear method of defining a complete dataset, we have the tools we need to have meaninful discussion about subsets of a dataset, and make progress on the all-or-nothing issue.
```
Comment (lightandluck) [inline text/0019-manifests.md]: typo: meaninful

File: text/0019-manifests.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+* The array of nodes MUST be sorted by number of descendants. 
+* When two nodes have the same number of descendants they MUST be sorted lexographically by node ID
+* The list of links MUST be sorted from low-to-high tuple values. eg: `[0,1]`, `[1,1]`, `[1,2]`.
+
+These requirements create a few nice properties:
+* The means that for DAGs with a single root node, the root ID will always be index 0 of `nodes`
+* Using this `nodes` organization, all leaf nodes will be grouped at the end of the list. 
+  * The element after highest index present in the `links` array is also the split point between leaf and branch nodes
+* supplying the same dag to the manifest function must be deterministic: `manifest_of_dag = manifest(dag)`
+  * Therfore: `hash(manifest_of_dag) == hash(manifest(dag))`
```
Comment (lightandluck) [inline text/0019-manifests.md]: typo: Therfore

File: text/0019-manifests.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+* In order to generate a manifest, you need the full DAG, or at least a trusted source of DAG IDs and links
+
+Manifests are intentionally limited in scope to make them easier to prove, faster to calculate, the list of nodes can be used as a base other structures can be built upon. by keeping manifests at a minimum they are easier to verify, forming a foundation for other.
```
Comment (lightandluck) [inline text/0019-manifests.md]: typo: capitalize 'by'

File: text/0019-manifests.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+A valid manifest has a few requirements:
+* The array of nodes MUST be sorted by number of descendants. 
+* When two nodes have the same number of descendants they MUST be sorted lexographically by node ID
+* The list of links MUST be sorted from low-to-high tuple values. eg: `[0,1]`, `[1,1]`, `[1,2]`.
+
+These requirements create a few nice properties:
+* The means that for DAGs with a single root node, the root ID will always be index 0 of `nodes`
+* Using this `nodes` organization, all leaf nodes will be grouped at the end of the list. 
+  * The element after highest index present in the `links` array is also the split point between leaf and branch nodes
+* supplying the same dag to the manifest function must be deterministic: `manifest_of_dag = manifest(dag)`
```
Comment (lightandluck) [inline text/0019-manifests.md]: Should 'dag' be all caps like other instances?

File: text/0019-manifests.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+Based on some light testing, if leaves have a max size of 256KB, a 100KB manifest encoded as CBOR can represent somewhere in the vicinity of 1GB of content. Having manifests that are roughly 1:10,000 in size makes them a reasonable choice for preflighting a DAG transfer over a network.
+
+I'd like to introduce two structures that build on Manifests: `DAGInfo` and `Completion`:
+
+### DAGInfo
+DAGInfo is intended to work like `os.FileInfo` for graph-based storage: a struct that describes important details about a graph. In terms of trues, the contents of DAGInfo should be considered gossip. DAGInfo's are *not* deterministic, and are intended to fill the role of application-specific needs when describing DAGs. 
+
+For Qri's purpos DAGInfo contains two fields: `sizes` and `paths`. `sizes` lists the size of each node in bytes, where each array index corresponds to the ID specified in `manifest.Nodes`. `sizes` is useful for a number of things, the most obivious being getting the the total size of the DAG by summing of all elements in the `sizes` array.
```
Comment (lightandluck) [inline text/0019-manifests.md]: typo: purpos, obivious

File: text/0019-manifests.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+* The means that for DAGs with a single root node, the root ID will always be index 0 of `nodes`
+* Using this `nodes` organization, all leaf nodes will be grouped at the end of the list. 
+  * The element after highest index present in the `links` array is also the split point between leaf and branch nodes
+* supplying the same dag to the manifest function must be deterministic: `manifest_of_dag = manifest(dag)`
+  * Therfore: `hash(manifest_of_dag) == hash(manifest(dag))`
+* In order to generate a manifest, you need the full DAG, or at least a trusted source of DAG IDs and links
+
+Manifests are intentionally limited in scope to make them easier to prove, faster to calculate, the list of nodes can be used as a base other structures can be built upon. by keeping manifests at a minimum they are easier to verify, forming a foundation for other.
+
+Based on some light testing, if leaves have a max size of 256KB, a 100KB manifest encoded as CBOR can represent somewhere in the vicinity of 1GB of content. Having manifests that are roughly 1:10,000 in size makes them a reasonable choice for preflighting a DAG transfer over a network.
```
Comment (lightandluck) [inline text/0019-manifests.md]: That's cool!

File: text/0019-manifests.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+  },
+  "sizes": [150,145,2340,256000,3404]
+}
+```
+
+Setting a `paths` property key `dataset.json` to `2` lets us leverage `links` to construct the subgraph starting at index 2 to construct the subraph at `dataset.json`.
+
+
+### Completion
+Completion tracks the presence of blocks described in a manifest. Completion can be used to store transfer progress, or be stored as a record of which blocks in a DAG are missing each element in the slice represents the index a block in a manifest.Nodes field, which contains the hash of a block needed to complete a manifest the element in the progress slice represents the transmission completion of that block's _locality_ (weather or not it's on your hard drive). Completion values must be a number from 0-100, where 0 = nothing locally, 100 = block is local.
```
Comment (lightandluck) [inline text/0019-manifests.md]: typo: need space between 'manifest.' and 'Nodes'

**Comment by b5** [comment]
Thanks so much for your read-through and review @lightandluck! Do you think it should be accepted?

**Comment by b5** [comment]
@ramfox you ok with this? I'd love to get it merged given that we've already written a bunch of code against this RFC

**Comment by b5** [comment]
Totally. Just wanted to see if you had an opinion on the matter, and if so we'd take it into consideration 😄. Thanks!

**Comment by b5** [comment]
Lovely. Thanks for the feedback all, merging!

**Pull request merged**